File: src/golang/axis-cli/data/AxisItem/AxisItem.go
package AxisItem

import (
	"axis-cli/data"
	"axis-cli/data/AxisUniverse"
	"time"

	"gorm.io/gorm"
)

type AxisItem struct {
	gorm.Model
	ID          uint64 `gorm:"primaryKey"`
	Title       string
	Description string
	TimeDue     time.Time
	Priority    uint8
	UniverseID  uint64
	Universe    AxisUniverse.AxisUniverse `gorm:"foreignKey:UniverseID"`
	Status      string
}

func Create(title string, description string, priority uint8, timeDue time.Time, universeName string, status string) error {
	// Create
	if priority == 0 {
		priority = 1
	}
	var universe AxisUniverse.AxisUniverse
	result := data.Db.First(&universe, "name = ?", universeName)
	if result.Error != nil {
		return result.Error
	}
	result = data.Db.Create(&AxisItem{Title: title, Description: description, Priority: priority, TimeDue: timeDue, Universe: universe, Status: status})
	return result.Error
}

func List() ([]AxisItem, error) {
	// Read
	var items []AxisItem
	result := data.Db.Model(&AxisItem{}).Limit(10).Find(&items)
	return items, result.Error
}

func ListToMap() ([]map[string]interface{}, error) {
	// Read
	var itemMaps []map[string]interface{}
	result := data.Db.Model(&AxisItem{}).Limit(10).Find(&itemMaps)
	return itemMaps, result.Error
}

func Find(id uint64) (AxisItem, error) {
	// Read
	var item AxisItem
	result := data.Db.First(&item, id)
	return item, result.Error
}

func FindToMap(id uint64) (map[string]interface{}, error) {
	// Read
	var itemMap map[string]interface{}
	result := data.Db.Model(&AxisItem{}).First(&itemMap, id)
	return itemMap, result.Error
}

func SearchToMap(keywords []string, searchDecription bool) ([]map[string]interface{}, error) {
	// Read
	var itemMaps []map[string]interface{}
	// TODO: enable searching description
	// TODO: enable searching multiple keywords
	result := data.Db.Model(&AxisItem{}).Where("title LIKE ?", "%"+keywords[0]+"%").Find(&itemMaps)
	return itemMaps, result.Error
}

func Update(id uint64, newTitle string, newDescription string, newPriority uint8, newTimeDue time.Time, newUniverseName string, newStatus string) error {
	var item AxisItem
	result := data.Db.First(&item, id)
	if newTitle != "" {
		item.Title = newTitle
	}
	if newDescription != "" {
		item.Description = newDescription
	}
	if newPriority != 0 {
		item.Priority = newPriority
	}
	if !newTimeDue.IsZero() {
		item.TimeDue = newTimeDue
	}
	if newUniverseName != "" {
		var universe AxisUniverse.AxisUniverse
		result := data.Db.First(&universe, "name = ?", newUniverseName)
		if result.Error != nil {
			return result.Error
		}
		item.UniverseID = universe.ID
	}
	if newStatus != "" {
		item.Status = newStatus
	}
	data.Db.Save(&item)
	return result.Error
}

func Delete(id uint64) error {
	var item AxisItem
	result := data.Db.First(&item, id)
	data.Db.Delete(&item)
	return result.Error
}

// func (u *AxisUniverse) BeforeDelete(tx *gorm.DB) (err error) {
// 	if u.Role == "admin" {
// 		return errors.New("admin user not allowed to delete")
// 	}
// 	return
// }

func init() {
	// Migrate the schema
	data.Db.AutoMigrate(&AxisItem{})
}

File: src/scala/axis-cli/src/main/scala/data/AxisItem.scala
package data

import slick.jdbc.H2Profile.api._

import java.time.OffsetDateTime
import scala.concurrent.Await
import scala.concurrent.duration.Duration


case class AxisItem(var id: Option[Int], var universeId: Option[Int], var title: String, var description: String = "", var timeDue: OffsetDateTime, var priority: Int, var status: String) {
    def toMap(): Map[String, Any] = {
        Map("id" -> id, "universeId" -> universeId, "title" -> title, "description" -> description, "timeDue" -> timeDue, "priority" -> priority, "status" -> status)
    }
}

class ItemEntity(tag: Tag) extends Table[AxisItem](tag, "Item") {
    def id = column[Int]("id", O.PrimaryKey, O.AutoInc)
    def universeId = column[Int]("universeId")
    def universe = foreignKey("universe", universeId, AxisUniverse.universes)(_.id, onUpdate=ForeignKeyAction.Restrict, onDelete=ForeignKeyAction.Cascade)
    def title = column[String]("title")
    def description = column[String]("description")
    def timeDue = column[OffsetDateTime]("timeDue")
    def priority = column[Int]("priority")
    def status = column[String]("status")
    def * = (id.?, universeId.?, title, description, timeDue, priority, status) <> ((AxisItem.apply _).tupled, AxisItem.unapply _)
}

object AxisItem {
    val items = TableQuery[ItemEntity]
    val printFields = Array("id", "universeId", "title", "description", "timeDue", "priority", "status")

    def create(title: String, universeName: String, description: String, timeDue: OffsetDateTime, priority: Int): Unit = {
        val universe = AxisUniverse.find(universeName)(0)
        var timeDueValid: OffsetDateTime = timeDue
        if (timeDue == null) {
            timeDueValid = OffsetDateTime.now()
        }
        var priorityValid: Int = priority
        if (priorityValid == 0) {
            priorityValid = 1
        }
        Await.result(DataUtils.db.run(
            items += AxisItem.apply(None, universe.id, title, description, timeDueValid, priorityValid, "New")
        ), Duration.Inf)
    }

    def list(): Seq[ItemEntity#TableElementType] = {
        Await.result(DataUtils.db.run(
            items.result
        ), Duration.Inf)
    }

    def search(keyword: String, searchDescription: Boolean): Seq[ItemEntity#TableElementType] = {
        var filter: ItemEntity => Rep[Boolean] = (e: ItemEntity) => false
        if (searchDescription) {
            filter = (e: ItemEntity) => (e.title.like(s"%${keyword}%") || e.description.like(s"%${keyword}%"))
        } else {
            filter = (e: ItemEntity) => (e.title.like(s"%${keyword}%"))
        }
        Await.result(DataUtils.db.run(
            items.filter(filter).result
        ), Duration.Inf)
    }

    def find(id: Int): Seq[ItemEntity#TableElementType] = {
        Await.result(DataUtils.db.run(
            items.filter(_.id === id).result
        ), Duration.Inf)
    }

    def delete(id: Int): Unit = {
        val affectedRowsCount = Await.result(DataUtils.db.run(
            items.filter(_.id === id).delete
        ), Duration.Inf)
        assert(affectedRowsCount > 0, s"Item with id: ${id} not found")
    }

    def update(id: Int, newTitle: String, newDescription: String, newTimeDue: OffsetDateTime, newPriority: Int, newStatus: String): Unit = {
        var entry = find(id)
        assert(entry.nonEmpty, s"Item with id: ${id} not found")

        var item = entry(0)
        if (newTitle != null && newTitle != "") {
            item.title = newTitle
        }
        if (newDescription != null && newDescription != "") {
            item.description = newDescription
        }
        if (newPriority != 0) {
            item.priority = newPriority
        }
        if (newStatus != null && newStatus != "") {
            item.status = newStatus
        }
        if (newTimeDue != null) {
            item.timeDue = newTimeDue
        }
        Await.result(DataUtils.db.run(
            items.update(item)
        ), Duration.Inf)
    }

    def move(id: Int, universeName: String) = {
        val uEntries = AxisUniverse.find(universeName)
        assert(uEntries.nonEmpty, s"Universe with name: ${universeName} not found")
        val universe = uEntries(0)

        val iEntries = find(id)
        assert(iEntries.nonEmpty, s"Item with id: ${id} not found")
        val item = iEntries(0)

        item.universeId = universe.id
        Await.result(DataUtils.db.run(
            items.update(item)
        ), Duration.Inf)
    }

}


